plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/AppEngineConfigurationUtil.java

```diff
@@ -0,0 +1,93 @@
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.Path;
+
+/**
+ * Utility class for resolving App Engine configuration files in a project. The goal of this class
+ * is to insulate callers from having to know how and when configuration files are placed within a
+ * project. For example, Java projects for the flexible environment typically place their YAML
+ * configuration files in {@code src/main/appengine} whereas Java projects for the standard
+ * environment place their XML configuration files within the {@code WEB-INF} directory.
+ *
+ * @see WebProjectUtil
+ */
+public class AppEngineConfigurationUtil {
+  /** A well-known location for App Engine configuration files. */
+  private static final IPath DEFAULT_CONFIGURATION_FILE_LOCATION = new Path("src/main/appengine");
+
+  /**
+   * Create an App Engine configuration file in the appropriate location for the project.
+   *
+   * @param project the hosting project
+   * @param relativePath the file name
+   * @param contents the content for the file
+   * @param overwrite if {@code true} then overwrite the file if it exists
+   */
+  public static IFile createConfigurationFile(
+      IProject project,
+      IPath relativePath,
+      InputStream contents,
+      boolean overwrite,
+      IProgressMonitor monitor)
+      throws CoreException {
+    IFolder appengineFolder = project.getFolder(DEFAULT_CONFIGURATION_FILE_LOCATION);
+    if (appengineFolder != null && appengineFolder.exists()) {
+      IFile destination = appengineFolder.getFile(relativePath);
+      if (!destination.exists()) {
+        IContainer parent = destination.getParent();
+        ResourceUtils.createFolders(parent, monitor);
+        destination.create(contents, true, monitor);
+      } else if (overwrite) {
+        destination.setContents(contents, IResource.FORCE, monitor);
+      }
+      return destination;
+    }
+    return WebProjectUtil.createFileInWebInf(project, relativePath, contents, overwrite, monitor);
+  }
+
+  /**
+   * Attempt to resolve the given file within the project's {@code WEB-INF}. Note that this method
+   * may return a file that is in a build location (e.g., {@code
+   * target/m2e-wtp/web-resources/WEB-INF}) which may be frequently removed or regenerated.
+   *
+   * @return the file location or {@code null} if not found
+   */
+  public static IFile findConfigurationFile(IProject project, IPath relativePath) {
+    IFolder appengineFolder = project.getFolder(DEFAULT_CONFIGURATION_FILE_LOCATION);
+    if (appengineFolder != null && appengineFolder.exists()) {
+      IFile destination = appengineFolder.getFile(relativePath);
+      if (destination.exists()) {
```

Comment:
No need to call `exists()` twice. This could basically be

```java
IFile destination = project.getFolder(DEFAULT_CONFIGURATION_FILE_LOCATION).getFile(relativePath);
```

Reply:
getFolder() returning null.

Reply:
Okay, that's fair. I still don't get the inconsistency (checking null for `getFolder()` but not for `getFile()`, etc.), but this is not a big deal.

Reply:
Point taken: I'll add a null check here too.

Reply:
You're right the null check shouldn't be necessary and I will instead fix up any tests.  

But back to your original question about combining into one chain: we need to split it up as we only create or update the file *providing* the `src/main/appengine` exists.  If it doesn't exist, then we put it into `WEB-INF`.